zokrates/gadgets/pedersenHasher.py

```diff
@@ -0,0 +1,217 @@
+    if len(name) > 28:
+        raise ValueError("Name too long")
+    data = b"%-28s%04X" % (name, i)
+    return Point.from_hash(data)
+
+
+def windows_to_dsl_array(windows):
+    bit_windows = (bitstring.BitArray(bin(i)).bin[::-1] for i in windows)
+    bit_windows_padded = ("{:0<3}".format(w) for w in bit_windows)
+    bitstr = "".join(bit_windows_padded)
+    # dsl = "[" + ", ".join(bitstr) + "]"
+    return list(bitstr)
+
+
+class PedersenHasher(object):
+    def __init__(self, name, segments=False):
+        self.name = name
+        if segments:
+            self.segments = segments
+            self.is_sized = True
+        else:
+            self.is_sized = False
+
+    def __gen_table(self):
+        name = self.name
+        segments = self.segments
+        assert (
+            self.is_sized == True
+        ), "Hasher size must be defined first, before lookup table can be created"
+        table = []
+        for j in range(0, segments):
+            if j % 62 == 0:
+                p = pedersen_hash_basepoint(name, j // 62)  # add to list
+            j = j % 62
+            if j != 0:
+                p = p.double().double().double().double()
+            # scalar = (window & 0b11) + 1
+            row = [p.mult(i + 1) for i in range(0, WINDOW_SIZE_BITS ** 2)]
+            table.append(row)
+
+        return table
+
+    def __hash_windows(self, windows, witness):
+        name = self.name
+
+        if self.is_sized == False:
+            self.segments = len(windows)
+            self.is_sized = True
+
+            generators = []
+            for j in range(0, self.segments):
+                if j % 62 == 0:
+                    current = pedersen_hash_basepoint(name, j // 62)  # add to list
+                j = j % 62
+                if j != 0:
+                    current = current.double().double().double().double()
+                generators.append(current)
+            self.generators = generators
+
+        segments = self.segments
+        assert (
+            len(windows) <= segments
+        ), "Number of windows exceeds pedersenHasher config. {} vs {}".format(
+            len(windows), segments
+        )
+        padding = (segments - len(windows)) * [0]  # pad to match number of segments
+        windows.extend(padding)
+        assert (
+            len(windows) == segments
+        ), "Number of windows does not match pedersenHasher config. {} vs {}".format(
+            len(windows), segments
+        )
+
+        # in witness mode return padded windows
+        if witness:
+            return windows_to_dsl_array(windows)
+
+        # TODO: define `62`,
+        # 248/62 == 4... ? CHUNKS_PER_BASE_POINT
+        result = Point.infinity()
+        for (g, window) in zip(self.generators, windows):
+            segment = g * ((window & 0b11) + 1)
+            if window > 0b11:
+                segment = segment.neg()
+            result += segment
+        return result
```

Comment:
I'm left with the impression that this method could re-use the `__gen_table` above to produce the lookup table and then select the correct entry (and potentially negate it) based on the 3-bit window. Might be worth a re-factor, if I'm not missing something.

Reply:
made the re-factor to reduce code doublication